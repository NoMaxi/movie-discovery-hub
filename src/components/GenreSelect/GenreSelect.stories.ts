import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import GenreSelect from "./GenreSelect";

const meta = {
    component: GenreSelect,
    args: {
        onSelect: fn(),
    },
} satisfies Meta<typeof GenreSelect>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        genres: ["All", "Comedy", "Horror", "Documentary"],
        selectedGenre: "Horror",
    },
};
