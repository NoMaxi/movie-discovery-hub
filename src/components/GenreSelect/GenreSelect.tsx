import { GenreFilter, Genre } from "@/types/common";
import { MAIN_GENRES, SECONDARY_GENRES } from "@/constants/constants";
import { GenreMoreDropdown } from "@/components/common/GenreMoreDropdown/GenreMoreDropdown";

interface GenreSelectProps {
    selectedGenre: GenreFilter;
    onSelect: (genre: GenreFilter) => void;
}

export const GenreSelect = ({ selectedGenre, onSelect }: GenreSelectProps) => {
    const isSecondaryGenre = selectedGenre !== "All" && SECONDARY_GENRES.includes(selectedGenre);

    return (
        <div className="flex flex-wrap items-center gap-x-8" data-testid="genre-select">
            {MAIN_GENRES.map((genre) => (
                <div
                    key={genre}
                    className={`
                        group relative px-2 py-1
                        text-lg font-medium uppercase cursor-pointer
                        transition-colors duration-200
                        ${genre === selectedGenre ? "text-text" : "text-gray-lighter hover:text-text"}
                    `}
                    onClick={() => onSelect(genre)}
                >
                    {genre}
                    <span
                        className={`
                            absolute left-0 h-[3px] w-0 -bottom-lg bg-primary
                            transition-all duration-300
                            ${genre === selectedGenre ? "w-full" : "group-hover:w-full"}
                        `}
                    ></span>
                </div>
            ))}

            <div className="relative group">
                <GenreMoreDropdown
                    options={SECONDARY_GENRES}
                    selectedGenre={isSecondaryGenre ? selectedGenre : undefined}
                    onSelectionChange={(genre: Genre) => onSelect(genre)}
                    className="genre-more-control"
                    buttonClassName={`${isSecondaryGenre ? "text-text" : "text-gray-lighter hover:text-text"}`}
                />
                <span
                    className={`
                        absolute left-0 h-[3px] w-0 -bottom-lg bg-primary
                        transition-all duration-300
                        ${isSecondaryGenre ? "w-full" : "group-hover:w-full"}
                    `}
                ></span>
            </div>
        </div>
    );
};
