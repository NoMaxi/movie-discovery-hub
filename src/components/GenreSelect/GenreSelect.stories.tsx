import { Meta, StoryFn, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import { GenreSelect } from "./GenreSelect";
import { GenreFilter } from "@/types/common";
import { useState } from "react";

const meta = {
    title: "Components/GenreSelect",
    component: GenreSelect,
    argTypes: {
        selectedGenre: {
            control: { type: "select" },
            options: [
                "All",
                "Drama",
                "Comedy",
                "Action",
                "Horror",
                "Adventure",
                "Animation",
                "Crime",
                "Documentary",
                "Family",
                "Fantasy",
                "History",
                "Music",
                "Mystery",
                "Romance",
                "Science Fiction",
                "Thriller",
                "TV Movie",
                "War",
                "Western",
            ],
            description: "Currently selected genre",
        },
        onSelect: {
            description: "Callback function triggered when a genre is selected",
        },
    },
    args: {
        onSelect: fn(),
    },
} satisfies Meta<typeof GenreSelect>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        selectedGenre: "Comedy",
    },
};

export const Interactive: StoryFn<typeof GenreSelect> = () => {
    const [selectedGenre, setSelectedGenre] = useState<GenreFilter>("All");

    return <GenreSelect selectedGenre={selectedGenre} onSelect={setSelectedGenre} />;
};
