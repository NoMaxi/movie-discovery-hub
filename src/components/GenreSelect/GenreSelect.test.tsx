import { render, screen } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { GenreSelect } from "./GenreSelect";

describe("GenreSelect", () => {
    let onSelectMock: jest.Mock;
    let user: UserEvent;

    beforeEach(() => {
        onSelectMock = jest.fn();
        user = userEvent.setup();
    });

    test("Should render main genre tabs and More dropdown", () => {
        const { asFragment } = render(<GenreSelect selectedGenre="Comedy" onSelect={onSelectMock} />);

        expect(screen.getByText("All")).toBeInTheDocument();
        expect(screen.getByText("Drama")).toBeInTheDocument();
        expect(screen.getByText("Comedy")).toBeInTheDocument();
        expect(screen.getByText("Action")).toBeInTheDocument();
        expect(screen.getByText("Horror")).toBeInTheDocument();
        expect(screen.getByText("More")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should highlight selected main genre", () => {
        const { asFragment } = render(<GenreSelect selectedGenre="Comedy" onSelect={onSelectMock} />);
        const selectedGenre = screen.getByText("Comedy");

        expect(selectedGenre).toHaveClass("text-[var(--color-text)]");
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should change styles correctly when different main genre is selected", () => {
        const { rerender } = render(<GenreSelect selectedGenre="Comedy" onSelect={onSelectMock} />);
        const initialSelectedGenre = screen.getByText("Comedy");

        expect(initialSelectedGenre).toHaveClass("text-[var(--color-text)]");

        rerender(<GenreSelect selectedGenre="Action" onSelect={onSelectMock} />);

        const newSelectedGenre = screen.getByText("Action");

        expect(initialSelectedGenre).not.toHaveClass("text-[var(--color-text)]");
        expect(initialSelectedGenre).toHaveClass("text-[var(--color-gray-lighter)]");
        expect(newSelectedGenre).toHaveClass("text-[var(--color-text)]");
    });

    test("Should call 'onSelect' prop with clicked main genre", async () => {
        render(<GenreSelect selectedGenre="Comedy" onSelect={onSelectMock} />);
        const clickedGenre = screen.getByText("Action");

        await user.click(clickedGenre);

        expect(onSelectMock).toHaveBeenCalledWith("Action");
    });

    test("Should highlight More dropdown when secondary genre is selected", () => {
        const { asFragment } = render(<GenreSelect selectedGenre="Romance" onSelect={onSelectMock} />);
        const moreButton = screen.getByText("Romance");

        expect(moreButton).toHaveClass("text-[var(--color-text)]");
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should call 'onSelect' prop when secondary genre is selected from dropdown", async () => {
        render(<GenreSelect selectedGenre="All" onSelect={onSelectMock} />);
        
        // Click on the "More" dropdown button to open it
        const moreButton = screen.getByText("More");
        await user.click(moreButton);
        
        // Find and click on a secondary genre option (e.g., "Adventure")
        const adventureOption = screen.getByText("Adventure");
        await user.click(adventureOption);
        
        // Verify that onSelect was called with the selected secondary genre
        expect(onSelectMock).toHaveBeenCalledWith("Adventure");
    });
});
