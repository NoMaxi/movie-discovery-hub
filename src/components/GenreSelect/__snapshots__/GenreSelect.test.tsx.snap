// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreSelect Should highlight More dropdown when secondary genre is selected 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap items-center gap-x-8"
    data-testid="genre-select"
  >
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      All
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Drama
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Comedy
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Action
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Horror
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="relative group"
    >
      <div
        class="genre-more-dropdown relative inline-block text-left genre-more-control"
        data-testid="genre-more-dropdown"
      >
        <button
          class="
                    genre-more-button inline-flex items-center justify-center gap-x-2 min-w-[80px]
                    px-2 py-1 text-lg font-medium uppercase
                    cursor-pointer transition-colors duration-200
                    focus:outline-none
                    
                        text-[var(--color-text)]
                    
                "
          type="button"
        >
          Romance
          <span
            class="transform transition-transform rotate-0"
          >
            <svg
              aria-hidden="true"
              class="w-3 h-3 fill-[var(--color-primary)] "
              data-testid="select-arrow-svg"
              viewBox="0 0 10 10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0 L10 0 L5 7 z"
                data-testid="select-arrow-path"
              />
            </svg>
          </span>
        </button>
      </div>
      <span
        class="
                        absolute left-0
                        h-[3px] w-0 -bottom-[16px]
                        bg-[var(--color-primary)]
                        transition-all duration-300
                        w-full
                    "
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`GenreSelect Should highlight selected main genre 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap items-center gap-x-8"
    data-testid="genre-select"
  >
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      All
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Drama
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-text)]
                    "
    >
      Comedy
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Action
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Horror
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="relative group"
    >
      <div
        class="genre-more-dropdown relative inline-block text-left genre-more-control"
        data-testid="genre-more-dropdown"
      >
        <button
          class="
                    genre-more-button inline-flex items-center justify-center gap-x-2 min-w-[80px]
                    px-2 py-1 text-lg font-medium uppercase
                    cursor-pointer transition-colors duration-200
                    focus:outline-none
                    
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    
                "
          type="button"
        >
          More Genres
          <span
            class="transform transition-transform rotate-0"
          >
            <svg
              aria-hidden="true"
              class="w-3 h-3 fill-[var(--color-primary)] "
              data-testid="select-arrow-svg"
              viewBox="0 0 10 10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0 L10 0 L5 7 z"
                data-testid="select-arrow-path"
              />
            </svg>
          </span>
        </button>
      </div>
      <span
        class="
                        absolute left-0
                        h-[3px] w-0 -bottom-[16px]
                        bg-[var(--color-primary)]
                        transition-all duration-300
                        group-hover:w-full
                    "
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`GenreSelect Should render main genre tabs and More dropdown 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap items-center gap-x-8"
    data-testid="genre-select"
  >
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      All
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Drama
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-text)]
                    "
    >
      Comedy
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Action
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="
                        group relative px-2 py-1
                        text-lg font-medium uppercase
                        cursor-pointer
                        transition-colors duration-200
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    "
    >
      Horror
      <span
        class="
                            className={
                            absolute left-0
                            h-[3px] w-0 -bottom-[16px]
                            bg-[var(--color-primary)]
                            transition-all duration-300
                            group-hover:w-full
                        "
      />
    </div>
    <div
      class="relative group"
    >
      <div
        class="genre-more-dropdown relative inline-block text-left genre-more-control"
        data-testid="genre-more-dropdown"
      >
        <button
          class="
                    genre-more-button inline-flex items-center justify-center gap-x-2 min-w-[80px]
                    px-2 py-1 text-lg font-medium uppercase
                    cursor-pointer transition-colors duration-200
                    focus:outline-none
                    
                        text-[var(--color-gray-lighter)] hover:text-[var(--color-text)]
                    
                "
          type="button"
        >
          More Genres
          <span
            class="transform transition-transform rotate-0"
          >
            <svg
              aria-hidden="true"
              class="w-3 h-3 fill-[var(--color-primary)] "
              data-testid="select-arrow-svg"
              viewBox="0 0 10 10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0 L10 0 L5 7 z"
                data-testid="select-arrow-path"
              />
            </svg>
          </span>
        </button>
      </div>
      <span
        class="
                        absolute left-0
                        h-[3px] w-0 -bottom-[16px]
                        bg-[var(--color-primary)]
                        transition-all duration-300
                        group-hover:w-full
                    "
      />
    </div>
  </div>
</DocumentFragment>
`;
